api/v1alpha1/contour_types.go

```diff
@@ -89,6 +96,49 @@ type NamespaceSpec struct {
 	RemoveOnDeletion bool `json:"removeOnDeletion,omitempty"`
 }
 
+// EndpointPublishing defines the schema to publish network endpoints and represents
+// the publishing type.
+type EndpointPublishing struct {
+	// Type is the type of publishing strategy to use. Valid values are:
+	//
+	// * LoadBalancerService
+	//
+	// Publishes endpoints using a Kubernetes LoadBalancer Service.
+	//
+	// In this configuration, the endpoints use container networking and a Kubernetes
+	// LoadBalancer Service is created.
+	//
+	// See: https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer
+	//
+	// * NodePortService
+	//
+	// Publishes the endpoints using a Kubernetes NodePort Service.
+	//
+	// In this configuration, the endpoints use container networking. A Kubernetes
+	// NodePort Service is created to publish the endpoints. The specific node ports
+	// are dynamically allocated by Kubernetes. To support static port allocations,
+	// user changes to the node port field of the managed NodePort Service will be
+	// preserved.
+	//
+	// See: https://kubernetes.io/docs/concepts/services-networking/service/#nodeport
+	//
+	// +kubebuilder:default=LoadBalancerService
+	Type EndpointPublishingType `json:"type,omitempty"`
+}
+
+// EndpointPublishingType is a way to publish network endpoints.
+// +kubebuilder:validation:Enum=LoadBalancerService;NodePortService
+type EndpointPublishingType string
+
+const (
+	// LoadBalancerService publishes an endpoint using a Kubernetes LoadBalancer
+	// Service.
+	LoadBalancerServiceStrategyType EndpointPublishingType = "LoadBalancerService"
+
+	// NodePortService publishes an endpoint using a Kubernetes NodePort Service.
+	NodePortServiceStrategyType EndpointPublishingType = "NodePortService"
```

Comment:
Maybe add `ClusterIP` for those who might need an internal ingress controller not exposed outside the cluster?

Reply:
@stevesloka I agree that `ClusterIP` or `ClusterIPService` will be a supported type. I would like to keep the initial API represented in this PR as minimal as possible and add types as needed. To provide parity with Contour's rendered manifest, this initial API needs to only support `LoadBalancerService` and `NodePortService`.